protocols/kad/src/handler.rs

```diff
@@ -772,20 +770,41 @@ where
             ));
         }
 
-        if let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
-            return Poll::Ready(event);
+        while let Poll::Ready(Some(event)) = self.outbound_substreams.poll_next_unpin(cx) {
+            if let ConnectionHandlerEvent::Custom(KademliaHandlerEvent::Available { substream }) =
+                event
+            {
+                self.reusable_outgoing_substreams.push(substream);
+            } else {
+                return Poll::Ready(event);
+            }
         }
 
         if let Poll::Ready(Some(event)) = self.inbound_substreams.poll_next_unpin(cx) {
             return Poll::Ready(event);
         }
 
-        let num_in_progress_outbound_substreams =
-            self.outbound_substreams.len() + self.num_requested_outbound_streams;
-        if num_in_progress_outbound_substreams < MAX_NUM_SUBSTREAMS {
-            if let Some(protocol) = self.requested_streams.pop_front() {
-                self.num_requested_outbound_streams += 1;
-                return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest { protocol });
+        while !self.pending_substream_requests.is_empty()
+            && self.outbound_substreams.len() + self.num_requested_outbound_streams
+                < MAX_NUM_SUBSTREAMS
+        {
+            if let Some((msg, user_data)) = self.pending_substream_requests.pop_front() {
```

Comment:
You could omit `pending_substream_requests` from the first condition here and instead `break` from the loop in an `else` block of the inner `if` (or change it to a `match` with a `break` on `None`).